src/main/java/org/threadly/concurrent/event/ListenerHelper.java

```diff
@@ -146,16 +145,17 @@ public void addListener(T listener, Executor executor) {
       if (addingFromCallingThread) {
         // we must create a new instance of listeners to prevent a ConcurrentModificationException
         // we know at this point that listeners can not be null
-        Map<T, Executor> newListeners = new HashMap<T, Executor>(listeners.size() + 1);
-        newListeners.putAll(listeners);
-        newListeners.put(listener, executor);
+        List<Pair<T, Executor>> newListeners = 
+            new ArrayList<Pair<T, Executor>>(listeners.size() + 1);
```

Comment:
this is a very rare condition here.  What we are doing here is the following:
* Check to see if we are adding a listener while being invoked inside another listener (ie a listener is being invoked, and that listener has the action of adding a listener.  We have similar logic elsewhere for removing too)
* If we are adding from inside a listener we hit this condition where we copy the contents, and add the listener to a new collection.  We must do this to prevent altering the collection we are currently iterating over, as doing so would cause a ConcurrentModificationException.
* In the other condition (ie adding a listener from anywhere else, which will be the case 99% of the time), we just add the listener directly to the ArrayList, and let the ArrayList expand as needed.

The reason I am only adding one here is because I suspect this condition to be very rare.  I also tend to assume that listener counts will be fairly low (say < 100).  I also am (without evidence) assuming that if you are hitting this condition, it may be a normal condition for you, and thus we may end up having to copy on every add, so it probably is not worth adding extra space into the collections.

Let me know if that makes sense to you or not.